File: Java/BOJ/12015 가장 긴 증가하는 부분 수열2.java
import java.io.*;
import java.util.*;
import java.math.*;

public class Main {
    public static void main(String[] args) throws IOException {
    	BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    	int N = Integer.parseInt(br.readLine());
    	StringTokenizer st = new StringTokenizer(br.readLine(), " ");
    	int[] A = new int[N];
    	ArrayList <Integer> result = new ArrayList<>();
    	result.add(0);
    	for (int i = 0; i < N; i ++) {
    		A[i] = Integer.parseInt(st.nextToken());
    	}
    	
    	for(int i : A) {
    		if (result.get(result.size()-1) < i) {
    			result.add(i);
    		}
    		else {
    			int left = 0;
    			int right = result.size();
    			while (left < right) {
    				int mid = (left + right)/2;
    				if (result.get(mid) < i) {
    					left = mid + 1;
    				}
    				else {
    					right = mid;
    				}
    			}
    			result.set(right, i);
    		}
    	}
    	System.out.println(result.size() - 1);
    }
}

File: Python/BOJ/12015 가장 긴 증가하는 부분 수열2.py
import sys
N = int(sys.stdin.readline())
A = list(map(int, sys.stdin.readline().split()))
result = [0]
for i in A: # 각 값에 대해
    if result[-1] < i: # 끝값보다 크면
        result.append(i) # 집어 넣고

    else: # 작으면
        left, right = 0, len(result)  #이분 탐색을 해보자
        while left < right:
            mid = (left + right)//2
            if result[mid] < i:
                left = mid + 1
            else :
                right = mid
        result[right] = i
print(len(result)-1)